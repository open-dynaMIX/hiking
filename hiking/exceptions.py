class HikingJsonLoaderException(Exception):
    pass


class HikingException(Exception):
    pass
