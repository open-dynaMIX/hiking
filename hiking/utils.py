import datetime
import logging
import os
from collections import namedtuple
from pathlib import Path
from typing import Optional, Union

from rich import box
from rich.console import Console

DATA_HOME = (
    Path(os.getenv("XDG_DATA_HOME")) / "hiking"
    if os.getenv("XDG_DATA_HOME")
    else Path.home() / ".local" / "share" / "hiking"
)

DB_PATH = DATA_HOME / "hikes.sqlite"
EDITOR = os.environ.get("EDITOR", "vi")
GPX_VIEWER = "/usr/bin/gpxsee"
DEFAULT_BOX_STYLE = box.HORIZONTALS
console = Console()
SlimDateRange = namedtuple(  # noqa: PYI024
    "SlimDateRange", ["lower", "upper"]
)

log_level = logging.WARNING


def setup_logging(debug: bool):
    global log_level  # noqa: PLW0603
    if debug:
        log_level = logging.DEBUG
    logging.basicConfig(format="%(levelname)-8s %(name)s:%(lineno)d %(message)s")
    logging.root.setLevel(log_level)
    logging.getLogger("sqlalchemy.engine").setLevel(
        logging.INFO if debug else logging.WARNING
    )


def pretty_timedelta(value: datetime.timedelta) -> str:
    hours = str(int(value.total_seconds() // 3600)).rjust(2, "0")
    minutes = str(int((value.total_seconds() // 60) % 60)).rjust(2, "0")
    return f"{hours}:{minutes}"


def format_value(
    value: Union[str, datetime.timedelta, datetime.date, float],
    attr: Optional[str] = None,
) -> str:
    if isinstance(value, datetime.timedelta):
        return pretty_timedelta(value)
    if isinstance(value, datetime.date):
        return str(value)
    if isinstance(value, str):
        return value

    assure_decimal_places = 0
    round_to_decimal_places = 2
    if attr and attr in ["elevation_gain", "elevation_loss"]:
        round_to_decimal_places = 0
    elif attr and attr == "distance":
        round_to_decimal_places = 1
        assure_decimal_places = 1
    elif attr and attr == "speed":
        assure_decimal_places = 2

    strs = str(round(value, round_to_decimal_places)).split(".")
    if len(strs) == 1:
        strs.append("")
    strs[1] = strs[1].rstrip("0")
    if len(strs[1]) < assure_decimal_places:
        strs[1] += "".ljust(assure_decimal_places - len(strs[1]), "0")
    return ".".join(strs) if strs[1] else strs[0]
