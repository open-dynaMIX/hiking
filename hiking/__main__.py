#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import logging
import sys

from hiking import commands
from hiking.arg_parsing import parse_arguments
from hiking.exceptions import HikingException, HikingJsonLoaderException
from hiking.import_export import JSON_IMPORT_EXAMPLE
from hiking.models import create_tables
from hiking.utils import DATA_HOME

logger = logging.getLogger(__name__)


def main():
    args = parse_arguments(sys.argv[1:])

    DATA_HOME.mkdir(exist_ok=True)
    create_tables()

    try:
        match args.command:
            case "show":
                commands.command_show(
                    args.ids,
                    args.daterange,
                    args.table_style,
                    args.order_key,
                    args.plot,
                )
            case "create" | "edit":
                commands.command_create_edit(pk=getattr(args, "id", None), gpx=args.gpx)
            case "delete":
                commands.command_delete(args.ids, args.all, args.force, args.quiet)
            case "import":
                commands.command_import(args.json_data)
            case "export":
                commands.command_export(
                    args.export_dir, args.ids, args.daterange, args.include_ids
                )

    except HikingJsonLoaderException as e:
        logger.warning(
            f"Invalid data in hiking.json: {e.args[0]}\n\nExpected format:\n{JSON_IMPORT_EXAMPLE}"
        )
    except HikingException as e:
        logger.error(f"Error: {e}")
    except (KeyboardInterrupt, EOFError):  # pragma: no cover
        sys.exit(0)


if __name__ == "__main__":  # pragma: no cover
    main()
