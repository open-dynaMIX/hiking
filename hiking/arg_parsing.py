import argparse
import datetime
import json
import os
from itertools import zip_longest
from pathlib import Path
from typing import List, Optional, Tuple

import gpxpy
from rich import box

from hiking.import_export import JSON_IMPORT_EXAMPLE
from hiking.models import Hike
from hiking.utils import DATA_HOME, DEFAULT_BOX_STYLE, SlimDateRange

# TODO: find a way to auto-detect this from rich
BOX_FORMATS = [
    "ASCII",
    "ASCII2",
    "ASCII_DOUBLE_HEAD",
    "SQUARE",
    "SQUARE_DOUBLE_HEAD",
    "MINIMAL",
    "MINIMAL_HEAVY_HEAD",
    "MINIMAL_DOUBLE_HEAD",
    "SIMPLE",
    "SIMPLE_HEAD",
    "SIMPLE_HEAVY",
    "HORIZONTALS",
    "ROUNDED",
    "HEAVY",
    "HEAVY_EDGE",
    "HEAVY_HEAD",
    "DOUBLE",
    "DOUBLE_EDGE",
    "MARKDOWN",
]


def get_valid_fields_for_args(exclude: Optional[List] = None):
    exclude = exclude or []
    return [
        field.info["name"]
        for field in Hike.FIELDS
        if field.info["data_view"] and field.info["name"] not in exclude
    ]


class DateRangeType:
    description = (
        "Only include hikes contained in provided daterange (default: all hikes)"
    )
    examples = (
        "Valid examples:\n"
        "1970-01-01\n"
        "1970-01-01/1970-02-01\n"
        "1970-01-01/ (all hikes from start date)\n"
        "/1970-01-01 (all hikes until end date)"
    )
    help = f"{description}\n{examples}"

    def __call__(self, raw: str, *args, **kwargs):
        start = end = raw
        splitted = raw.split("/")
        if len(splitted) == 2 and all(splitted):
            start, end = splitted
        elif start.endswith("/"):
            start = start.rstrip("/")
            end = None
        elif end.startswith("/"):
            end = end.lstrip("/")
            start = None

        try:
            start = (
                datetime.datetime.strptime(start, "%Y-%m-%d").date()
                if start
                else datetime.date.min
            )
            end = (
                datetime.datetime.strptime(end, "%Y-%m-%d").date()
                if end
                else datetime.date.max
            )
        except ValueError as e:
            raise argparse.ArgumentTypeError(f"{e.args[0]}\n{self.examples}")
        return SlimDateRange(start, end)


class WritableDirPathType:
    def __call__(self, raw: str, *args, **kwargs):
        directory = Path(raw)
        if (
            not directory.exists()
            or not directory.is_dir()
            or not os.access(directory, os.W_OK)
        ):
            raise argparse.ArgumentTypeError(
                f'Cannot write to directory: "{directory.absolute()}". '
                f"Make sure it exists and is writable."
            )
        return directory


class GPXFileType(argparse.FileType):
    def __call__(self, *args, **kwargs):
        file = super().__call__(*args, **kwargs)
        try:
            gpx_xml = file.read()
            gpxpy.parse(gpx_xml)
        except Exception as e:
            raise argparse.ArgumentTypeError(f"Cannot read *.gpx file: {str(e)}")
        return gpx_xml


class JsonFileType(argparse.FileType):
    def __call__(self, *args, **kwargs):
        file = super().__call__(*args, **kwargs)
        try:
            data = json.load(file)
        except Exception as e:
            raise argparse.ArgumentTypeError(f"Cannot read *.json file: {str(e)}")
        return data


def validate_order_key(value: str) -> Tuple[str, bool]:
    reverse = False
    if value.startswith("-"):
        reverse = True
        value = value.lstrip("-")
    if value not in get_valid_fields_for_args():
        raise argparse.ArgumentTypeError("Invalid order_key")
    return value, reverse


def validate_plot(value: str) -> Tuple[str, str]:
    try:
        x, y = tuple(value.split(","))
        for i in x, y:
            assert i in get_valid_fields_for_args(["name"])
        return x, y
    except (ValueError, AssertionError):
        raise argparse.ArgumentTypeError("plot")


def validate_table_style(value: str) -> Tuple[str, str]:
    try:
        box_style = getattr(box, value.upper())
    except AttributeError:
        raise argparse.ArgumentTypeError("Invalid table-style")
    return box_style


def set_default_subparser(
    parser: argparse.ArgumentParser, default_subcommand: str, raw_args: List[str]
):
    subparser_found = False
    for arg in raw_args:
        if arg in ["-h", "--help"]:  # pragma: no cover
            break
    else:
        for x in parser._subparsers._actions:
            if not isinstance(x, argparse._SubParsersAction):
                continue
            for sp_name in x._name_parser_map.keys():
                if sp_name in raw_args:
                    subparser_found = True
        if not subparser_found:
            raw_args.insert(0, default_subcommand)


def parse_arguments(raw_args: List[str]) -> argparse.Namespace:
    """
    Parse all arguments.
    """

    def format_list(data: List[str]):
        data.sort()
        col_1 = data[: int(round(len(data) / 2))]
        col_2 = data[int(round(len(data) / 2)) :]
        table_data = list(zip_longest(col_1, col_2, fillvalue=""))
        longest = 0
        for i in table_data:
            if len(i[0]) > longest:
                longest = len(i[0])
        result = [f"{i[0].ljust(longest)}{' ' * 5}{i[1]}" for i in table_data]
        return "\n".join(result)

    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawTextHelpFormatter, prog="hiking"
    )

    debug_arg_dict = {
        "help": "Show debug information (log queries)",
        "action": "store_true",
    }

    subparsers = parser.add_subparsers(dest="command")

    show = subparsers.add_parser(
        "show",
        help="Show hike(s) (default)",
        description="Show hike(s) (default)",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    show.add_argument(
        "ids",
        metavar="ID",
        help="Hike ID",
        nargs="*",
        type=int,
    )

    show.add_argument(
        "-d",
        "--daterange",
        help=DateRangeType.help,
        type=DateRangeType(),
        default=SlimDateRange(datetime.date.min, datetime.date.max),
    )

    show.add_argument(
        "-s",
        "--search",
        help="Search for text in name and body (case insensitive)",
        type=str,
    )

    show.add_argument(
        "-t",
        "--table-style",
        help=(
            "Table format style (default: simple)\n"
            f"Available options:\n{format_list(BOX_FORMATS)}"
        ),
        default=DEFAULT_BOX_STYLE,
        type=validate_table_style,
    )

    show.add_argument(
        "-o",
        "--order-key",
        help=(
            'Key to use for hike sorting. To reverse, prepend with "-".\n'
            "Available options:\n"
            f"{format_list(get_valid_fields_for_args())}"
        ),
        default=("date", False),
        type=validate_order_key,
    )

    show.add_argument(
        "--plot",
        help=(
            "Fields to plot in a graph.\n"
            "*experimental*\n"
            "Example:\n"
            '"date,distance"\n'
            "Available options:\n"
            f"{format_list(get_valid_fields_for_args(exclude=['name']))}"
        ),
        default=(),
        type=validate_plot,
    )

    show.add_argument("--debug", **debug_arg_dict)

    create = subparsers.add_parser(
        "create",
        help="Create a new record",
        description="Create a new record.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    create.add_argument(
        "--gpx",
        metavar="GPX_FILE",
        type=GPXFileType("r"),
        help="Import from *.gpx-file",
    )

    create.add_argument("--debug", **debug_arg_dict)

    edit = subparsers.add_parser(
        "edit",
        help="Edit a record",
        description="Edit a record.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    edit.add_argument(
        "id",
        metavar="ID",
        help="Hike ID",
        type=int,
    )

    edit.add_argument(
        "--gpx",
        metavar="GPX_FILE",
        type=GPXFileType("r"),
        help="Import from *.gpx-file",
    )

    edit.add_argument("--debug", **debug_arg_dict)

    delete = subparsers.add_parser(
        "delete",
        help="Delete records by ID",
        description="Delete records by ID.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    delete.add_argument(
        "-f",
        "--force",
        help="Do not ask before deletion",
        action="store_true",
    )

    delete.add_argument(
        "-q",
        "--quiet",
        help="Do not display hikes before deletion",
        action="store_true",
    )

    delete.add_argument(
        "-a",
        "--all",
        help="Delete all hikes",
        action="store_true",
    )

    delete.add_argument(
        "ids",
        metavar="ID",
        help="Hike ID",
        nargs="*",
        type=int,
    )

    delete.add_argument("--debug", **debug_arg_dict)

    _import = subparsers.add_parser(
        "import",
        help="Import records from JSON",
        description=f"Import records from JSON.\nFormat:\n{JSON_IMPORT_EXAMPLE}",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    _import.add_argument(
        "json_data",
        metavar="JSON_FILE",
        help="Path to JSON file",
        type=JsonFileType("r"),
    )

    _import.add_argument("--debug", **debug_arg_dict)

    export = subparsers.add_parser(
        "export",
        help="Export records as JSON and GPX",
        description="Export records as JSON and GPX.",
        formatter_class=argparse.RawTextHelpFormatter,
    )

    export.add_argument(
        "export_dir",
        metavar="EXPORT_DIR",
        help="Path to export directory",
        type=WritableDirPathType(),
    )

    export.add_argument(
        "ids",
        metavar="ID",
        help="Hike ID",
        nargs="*",
        type=int,
    )

    export.add_argument(
        "-d",
        "--daterange",
        help=DateRangeType.help,
        type=DateRangeType(),
        default=SlimDateRange(datetime.date.min, datetime.date.max),
    )

    export.add_argument(
        "-i",
        "--include-ids",
        help='Include IDs in export. Needed for "update", must be omitted for "create"',
        action="store_true",
    )

    export.add_argument("--debug", **debug_arg_dict)

    set_default_subparser(parser, "show", raw_args)

    args = parser.parse_args(raw_args)

    if args.command == "delete" and not args.ids and not args.all:
        raise parser.error("IDs or --all must be provided")
    elif args.command == "delete" and args.ids and args.all:
        raise parser.error("Ambiguous argument: IDs and --all provided")

    try:
        WritableDirPathType()(DATA_HOME.parent)
    except argparse.ArgumentTypeError:
        raise parser.error(f"Cannot write to user data director: {DATA_HOME.parent}")

    return args
