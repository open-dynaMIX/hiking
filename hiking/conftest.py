from pathlib import Path

import pytest
from pytest_factoryboy.fixture import register

from hiking import factories
from hiking.models import create_tables, engine, session

OWN_DIR = Path(__file__).resolve().parent
GPX_TEST_FILE = OWN_DIR / "tests" / "data" / "test.gpx"


register(factories.HikeFactory)


@pytest.fixture(scope="session", autouse=True)
def setup_db():
    create_tables()


@pytest.fixture(autouse=True)
def db():
    yield
    with engine.connect() as con:
        con.execute("DELETE FROM hikes;")
    session.commit()
    session.close()
    session.begin()


@pytest.fixture
def gpx_xml():
    with GPX_TEST_FILE.open("r") as f:
        xml = f.read()
    return xml
