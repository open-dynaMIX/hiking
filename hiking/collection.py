import datetime
from dataclasses import dataclass
from typing import List, Tuple, Union

from sqlalchemy import Interval, func
from sqlalchemy.orm import Query

from hiking.db_utils import session
from hiking.models import Hike, get_filtered_query
from hiking.utils import SlimDateRange, format_value


@dataclass
class HikeCollection:
    hikes: Query

    def get_hikes_attr_list(
        self, attr: str
    ) -> List[Union[float, int, datetime.date, datetime.timedelta]]:
        return [getattr(hike, attr) for hike in self.hikes.all()]

    def sum(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        is_interval = isinstance(getattr(Hike, attr).expression.type, Interval)
        if Hike.FIELD_PROPS[attr]["calculated_value"] or is_interval:
            attr_list = self.get_hikes_attr_list(attr)
            if is_interval:
                return sum(attr_list, datetime.timedelta())

            return sum(attr_list)  # pragma: no cover

        result = session.query(func.sum(getattr(Hike, attr))).first()[0]
        return result

    def avg(self, attr: str = "distance") -> Union[float, datetime.timedelta]:
        if attr == "duration":
            return (
                sum([h.duration for h in self.hikes.all()], datetime.timedelta())
                / self.hikes.count()
            )
        return sum(getattr(h, attr) for h in self.hikes.all()) / self.hikes.count()

    def max(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        if attr == "speed":
            return sorted(self.hikes.all(), key=lambda x: x.speed, reverse=True)[
                0
            ].speed

        result = getattr(
            self.hikes.order_by(getattr(Hike, attr).desc()).limit(1).first(), attr
        )
        return result

    def min(
        self, attr: str = "distance"
    ) -> Union[float, int, datetime.date, datetime.timedelta]:
        if attr == "speed":
            return sorted(self.hikes.all(), key=lambda x: x.speed)[0].speed

        result = getattr(
            self.hikes.order_by(getattr(Hike, attr).asc()).limit(1).first(), attr
        )
        return result

    def get_hikes_stats(self, order_params: Tuple[str, bool]) -> List[List[str]]:
        if order_params[0] == "speed":
            hike_list = sorted(
                self.hikes.all(), key=lambda x: x.speed, reverse=order_params[1]
            )
            return [hike.get_stats() for hike in hike_list]

        order = getattr(Hike, order_params[0])
        if order_params[1]:
            order = order.desc()
        query = self.hikes.order_by(order)
        return [hike.get_stats() for hike in query]

    def calc_and_format_value(self, calc: str, attr: str) -> str:
        result = getattr(self, calc)(attr)
        return format_value(result, attr)

    def get_totals(self) -> List[str]:
        def get_summary_cell(attr: str, supported_calculations: List[str]):
            if supported_calculations:
                cell = {}
                for calc, pretty_calc in [
                    ("sum", "Σ "),
                    ("avg", "⌀ "),
                    ("max", "↑ "),
                    ("min", "↓ "),
                ]:
                    cell[pretty_calc] = "-"
                    if calc in supported_calculations:
                        cell[pretty_calc] = self.calc_and_format_value(calc, attr)
                return cell

        d = [
            "",
            "STATS",
            str(self.hikes.count()),
            *[
                get_summary_cell(attr, config["supported_calculations"])
                for attr, config in Hike.FIELD_PROPS.items()
                if config["supported_calculations"]
            ],
        ]
        return d

    def get_collection_stats(
        self, order_params: Tuple[str, bool], add_totals: bool = True
    ) -> Tuple[List, List]:
        stats = self.get_hikes_stats(order_params)

        result = [
            *stats,
        ]

        footer = None
        if add_totals and len(stats) > 1:
            # Only add totals if more than one hike is present
            footer = self.get_totals()

        return result, footer

    def __str__(self) -> str:
        return (
            f"<HikeCollection - "
            f"containing {self.hikes.count()} hikes - "
            f"total {self.sum()}km>"
        )

    def __repr__(self) -> str:
        return self.__str__()


def get_collection(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = get_filtered_query(ids, daterange)
    collection = HikeCollection(hikes=query)
    return collection
