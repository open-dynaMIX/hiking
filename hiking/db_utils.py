from typing import List

from hiking.collection import HikeCollection
from hiking.models import Hike, session
from hiking.utils import SlimDateRange


def get_filtered_query(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = (
        session.query(Hike)
        .filter(Hike.date >= daterange.lower)
        .filter(Hike.date <= daterange.upper)
    )
    if ids:
        query = query.filter(Hike.id.in_(ids))
    return query


def get_collection(
    ids: List[int],
    daterange: "SlimDateRange",
):
    query = get_filtered_query(ids, daterange)
    collection = HikeCollection(hikes=query, session=session)
    return collection
