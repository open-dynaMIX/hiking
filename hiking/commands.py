from pathlib import Path
from typing import List, Optional, Tuple

from rich import box
from rich.markdown import Markdown
from rich.prompt import Confirm
from rich.table import Table

from hiking.collection import HikeCollection
from hiking.db_utils import session
from hiking.exceptions import HikingException
from hiking.gpx import get_elevation_profile
from hiking.import_export import json_exporter, json_importer
from hiking.interactivity import display_gpx, user_create_edit_interaction
from hiking.models import Hike, get_filtered_query
from hiking.plot import plot
from hiking.utils import DEFAULT_BOX_STYLE, SlimDateRange, console


def get_table(
    collection: HikeCollection,
    order_params: Tuple[str, bool],
    table_style: box.Box = box.SIMPLE,
    add_totals: bool = True,
) -> Table:
    data, footer = collection.get_collection_stats(order_params, add_totals=add_totals)
    headers = {attr: config["pretty_name"] for attr, config in Hike.FIELD_PROPS.items()}

    arrow = "▲" if order_params[1] else "▼"
    headers[order_params[0]] = f"{headers[order_params[0]]} {arrow}"

    table = Table(
        title="Hikes",
        show_header=True,
        show_footer=bool(footer),
        header_style="bold",
        box=table_style,
    )

    def get_grid(content: dict):
        if isinstance(content, str):
            return content

        grid = Table.grid(expand=True)
        grid.add_column()
        grid.add_column(justify="right")
        for k, v in content.items():
            grid.add_row(k, v)
        return grid

    table.add_column("ID", justify="right", footer=footer and get_grid(footer[0]))
    table.add_column("Date", justify="left", footer=footer and get_grid(footer[1]))
    table.add_column("Name", justify="left", footer=footer and get_grid(footer[2]))
    table.add_column("➡ km", justify="right", footer=footer and get_grid(footer[3]))
    table.add_column("⬈ m", justify="right", footer=footer and get_grid(footer[4]))
    table.add_column("⬊ m", justify="right", footer=footer and get_grid(footer[5]))
    table.add_column("⏱", justify="right", footer=footer and get_grid(footer[6]))
    table.add_column("km/h", justify="right", footer=footer and get_grid(footer[7]))

    for d in data:
        table.add_row(*d)

    return table


def draw_plot(
    collection: HikeCollection, x_attr: str = "date", y_attr: str = "distance"
):
    x = collection.get_hikes_attr_list(x_attr)
    y = collection.get_hikes_attr_list(y_attr)
    return plot(
        x=x,
        y=y,
        xlabel=Hike.FIELD_PROPS[x_attr]["pretty_name"],
        ylabel=Hike.FIELD_PROPS[y_attr]["pretty_name"],
        x_limit_min=1 if not x_attr == "date" else None,
    )


def command_create_edit(pk: int = None, gpx: str = None):
    """
    Create or edit a record.

    If `pk` is provided, `edit` will be performed.
    """
    hike = session.query(Hike).get(pk) if pk is not None else Hike()

    if not hike:
        raise HikingException("No hike found with provided ID")

    if gpx:
        hike.load_gpx(gpx)

    user_create_edit_interaction(hike)

    stats = hike.get_detail_stats()
    stats["GPX"] = "available" if hike.gpx_xml else "not available"

    if not pk:
        stats.pop("ID")

    print_detail_stats(stats)

    confirmation = Confirm.ask("Should this hike be written to the DB?")

    if not confirmation:  # pragma: no cover
        print("Aborting")
        return

    hike.save()


def command_delete(ids: List[int], delete_all: bool, force: bool, quiet: bool):
    collection = HikeCollection(
        hikes=get_filtered_query(ids=None if delete_all else ids, daterange=None)
    )

    if not collection.hikes.first():
        raise HikingException("No hikes found with provided ID(s)")
    elif ids and collection.hikes.count() < len(ids):
        raise HikingException("Invalid ID(s) provided")

    if not quiet:
        console.print("This action will delete following hikes:\n")
        table = get_table(
            collection,
            ("date", False),
            add_totals=False,
        )
        console.print(table)

    if not force:
        confirmation = Confirm.ask("Are you sure you want them deleted?")

        if not confirmation:
            print("Aborting")
            return

    for hike in collection.hikes:
        hike.delete()


def command_import(json_data: List[dict]):
    json_importer(json_data)


def command_export(
    export_dir: Path, ids: List[int], daterange: "SlimDateRange", include_ids: bool
):
    query = get_filtered_query(ids=ids, daterange=daterange, load_all_columns=True)
    json_exporter(query, export_dir, include_ids)


def print_detail_stats(stats: dict, table_style: box = DEFAULT_BOX_STYLE):
    title = stats.pop("Name")
    table = Table(box=table_style, show_header=False, title=title, min_width=40)
    for k, v in stats.items():
        table.add_row(k, str(v))
    console.print(table)


def detail_view(collection: HikeCollection):
    hike = collection.hikes.first()
    stats = hike.get_detail_stats()

    print_detail_stats(stats)
    console.print()

    md = Markdown(hike.body or "")
    if md:
        console.print(md)
        console.print()

    print(get_elevation_profile(hike))

    console.print()
    if hike.gpx_xml:
        display_gpx(hike.gpx_xml)


def command_show(
    ids: List[int],
    daterange: "SlimDateRange",
    table_style: box,
    order_params: Tuple[str, bool],
    plot_params: Tuple[Optional[str], Optional[str]],
) -> None:
    if not session.query(Hike).first():
        raise HikingException(
            'No hikes in DB. Add some hikes with "create" or "import"'
        )

    collection = HikeCollection(hikes=get_filtered_query(ids, daterange))
    if not collection.hikes.first():
        raise HikingException("No hikes found with given parameters")

    if len(ids) != 1:
        table = get_table(collection, order_params, table_style)
        console.print(table)

    if len(ids) == 1:
        detail_view(collection)

    if plot_params:
        plot = draw_plot(collection, plot_params[0], plot_params[1])
        print(plot)
