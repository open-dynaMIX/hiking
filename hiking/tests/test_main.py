import sys
from collections import namedtuple

import pytest

from hiking import __main__, commands
from hiking.__main__ import main
from hiking.tests.utils import ansi_escape

ArgsMock = namedtuple(
    "args",
    [
        "command",
        "id",
        "ids",
        "all",
        "force",
        "quiet",
        "json_data",
        "daterange",
        "table_style",
        "order_key",
        "plot",
        "gpx",
        "export_dir",
        "include_ids",
    ],
    defaults=(None,) * 13,
)


def test_main_end_to_end(snapshot, capsys, sys_argv, collection):
    sys.argv = ["tests", "show", "--order-key=-distance"]
    main()
    captured = capsys.readouterr()[0]
    assert ansi_escape(captured) == snapshot


@pytest.mark.parametrize(
    "command, args",
    [
        ("command_create_edit", ["create"]),
        ("command_delete", ["delete"]),
        ("command_import", ["import"]),
        ("command_export", ["export", "/tmp/"]),
    ],
)
def test_main_commands(mocker, sys_argv, command, args):
    mocker.patch.object(commands, command)
    mocked_args = ArgsMock(command=args[0])
    mocker.patch.object(
        __main__,
        "parse_arguments",
        return_value=mocked_args,
    )
    sys.argv = ["tests", *args]
    main()


def test_main_hiking_json_loader_exception(mocker, caplog, snapshot):
    args = ArgsMock(command="import", json_data=[{"foo": "bar"}])
    mocker.patch.object(
        __main__,
        "parse_arguments",
        return_value=args,
    )
    main()
    assert len(caplog.records) == 1
    assert caplog.records[0].msg == snapshot


def test_main_hiking_exception(mocker, caplog, snapshot):
    args = ArgsMock(command="delete", ids=[23])
    mocker.patch.object(
        __main__,
        "parse_arguments",
        return_value=args,
    )
    main()
    assert len(caplog.records) == 1
    assert caplog.records[0].msg == snapshot
