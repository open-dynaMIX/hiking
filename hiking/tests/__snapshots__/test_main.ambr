# serializer version: 1
# name: test_main_end_to_end[False]
  '''
                                       Hikes                                      
   ────────────────────────────────────────────────────────────────────────────── 
    ID   Date       Name       GPX     ➡ km      ⬈ m      ⬊ m         ⏱     km/h  
   ────────────────────────────────────────────────────────────────────────────── 
     2   2016-05…   Gina         ✗     17.8     2732     3346     03:28     5.12  
                    Lynch                                                         
     1   2023-06…   Paul         ✗     17.4     3268     2652     03:50     4.53  
                    Diaz                                                          
     3   2025-03…   John         ✗     14.9     2452      733     03:17     4.55  
                    Rhodes                                                        
   ────────────────────────────────────────────────────────────────────────────── 
         STATS      3                Σ 50.2   Σ 8452   Σ 6731   Σ 10:36   Σ    -  
                                     ⌀ 16.7   ⌀ 2817   ⌀ 2244   ⌀ 03:32   ⌀ 4.74  
                                     ↑ 17.8   ↑ 3268   ↑ 3346   ↑ 03:50   ↑ 5.12  
                                     ↓ 14.9   ↓ 2452   ↓  733   ↓ 03:17   ↓ 4.53  
   ────────────────────────────────────────────────────────────────────────────── 
  
  '''
# ---
# name: test_main_end_to_end[True]
  '''
                                       Hikes                                      
   ────────────────────────────────────────────────────────────────────────────── 
    ID   Date       Name       GPX     ➡ km      ⬈ m      ⬊ m         ⏱     km/h  
   ────────────────────────────────────────────────────────────────────────────── 
     2   2016-05…   Gina         ✗     17.8     2732     3346     03:28     5.12  
                    Lynch                                                         
     1   2023-06…   Paul         ✗     17.4     3268     2652     03:50     4.53  
                    Diaz                                                          
     3   2025-03…   John         ✗     14.9     2452      733     03:17     4.55  
                    Rhodes                                                        
   ────────────────────────────────────────────────────────────────────────────── 
         STATS      3                Σ 50.2   Σ 8452   Σ 6731   Σ 10:36   Σ    -  
                                     ⌀ 16.7   ⌀ 2817   ⌀ 2244   ⌀ 03:32   ⌀ 4.74  
                                     ↑ 17.8   ↑ 3268   ↑ 3346   ↑ 03:50   ↑ 5.12  
                                     ↓ 14.9   ↓ 2452   ↓  733   ↓ 03:17   ↓ 4.53  
   ────────────────────────────────────────────────────────────────────────────── 
  
  '''
# ---
# name: test_main_hiking_exception
  'Error: No hikes found with provided ID(s)'
# ---
# name: test_main_hiking_json_loader_exception
  '''
  Invalid data in hiking.json: Invalid JSON data:
  Missing fields: date, distance, duration, elevation_gain, elevation_loss, gpx_file, name
  Unknown fields: foo
  
  Expected format:
  [
      {
          "id": "$Integer (optional; update if present)",
          "name": "$String",
          "body": "$String",
          "date": "YYY-MM-DD",
          "distance": "$Foat",
          "elevation_gain": "$Integer",
          "elevation_loss": "$Integer",
          "duration": "$Integer",
          "gpx": "$String (path to file; optional)"
      }
  ]
  '''
# ---
